File: serverless/update.py
import json
import os
import base64
import pymysql


def _headers():
    return {
        "Content-Type": "application/json",
        "Access-Control-Allow-Origin": "*",
        "Access-Control-Allow-Methods": "GET,POST,PUT,DELETE,OPTIONS",
        "Access-Control-Allow-Headers": "Content-Type,Authorization",
    }


def _conn():
    return pymysql.connect(
        host=os.environ["DB_HOST"],
        port=int(os.environ.get("DB_PORT", "3306")),
        user=os.environ["DB_USER"],
        password=os.environ["DB_PASSWORD"],
        database=os.environ["DB_NAME"],
        charset="utf8mb4",
        cursorclass=pymysql.cursors.DictCursor,
        autocommit=True,
    )


def _parse_body(event):
    body = event.get("body")
    if not body:
        return {}
    if event.get("isBase64Encoded"):
        body = base64.b64decode(body).decode("utf-8")
    try:
        return json.loads(body)
    except Exception:
        return {}


def _resolve_id(event):
    path_params = event.get("pathParameters") or {}
    query_params = event.get("queryStringParameters") or {}
    if "id" in path_params:
        try:
            return int(path_params["id"])
        except Exception:
            return None
    if "id" in query_params:
        try:
            return int(query_params["id"])
        except Exception:
            return None
    return None


def lambda_handler(event, context):
    if event.get("httpMethod") == "OPTIONS":
        return {"statusCode": 204, "headers": _headers(), "body": ""}

    try:
        _id = _resolve_id(event)
        if _id is None:
            return {
                "statusCode": 400,
                "headers": _headers(),
                "body": json.dumps({"message": "Missing or invalid 'id' parameter."}),
            }

        data = _parse_body(event)
        title = data.get("Title") if "Title" in data or "title" in data else None
        if title is None and "title" in data:
            title = data.get("title")
        description = (
            data.get("Description") if "Description" in data or "description" in data else None
        )
        if description is None and "description" in data:
            description = data.get("description")
        status = data.get("Status") if "Status" in data or "status" in data else None
        if status is None and "status" in data:
            status = data.get("status")

        set_parts = []
        params = []
        if title is not None:
            set_parts.append("Title = %s")
            params.append(title)
        if description is not None:
            set_parts.append("Description = %s")
            params.append(description)
        if status is not None:
            set_parts.append("Status = %s")
            params.append(status)

        if not set_parts:
            return {
                "statusCode": 400,
                "headers": _headers(),
                "body": json.dumps({"message": "No fields to update."}),
            }

        with _conn() as conn:
            with conn.cursor() as cur:
                sql = f"UPDATE Tasks SET {', '.join(set_parts)} WHERE Id = %s"
                cur.execute(sql, (*params, _id))

            with conn.cursor() as cur:
                cur.execute(
                    "SELECT Id, Title, Description, Status FROM Tasks WHERE Id = %s",
                    (_id,),
                )
                row = cur.fetchone()

        if not row:
            return {
                "statusCode": 404,
                "headers": _headers(),
                "body": json.dumps({"message": "Not found."}),
            }

        return {"statusCode": 200, "headers": _headers(), "body": json.dumps(row)}

    except Exception as e:
        print(f"ERROR: {e}")
        return {
            "statusCode": 500,
            "headers": _headers(),
            "body": json.dumps({"message": "Internal Server Error"}),
        }

File: Serverless-CRUD/Update.cs
using System.Text.Json;
using Amazon.Lambda.APIGatewayEvents;
using Amazon.Lambda.Core;
using MySqlConnector;
using System.Threading.Tasks;
using System;
using System.Collections.Generic;
using System.Text;

namespace ServerlessCrud
{
    public class Update
    {
        public async Task<APIGatewayProxyResponse> FunctionHandler(APIGatewayProxyRequest request, ILambdaContext context)
        {
            try
            {
                var id = ResolveId(request);
                if (id == null)
                {
                    return JsonResponse(400, new { message = "Missing or invalid 'id' parameter." });
                }

                if (string.IsNullOrWhiteSpace(request?.Body))
                {
                    return JsonResponse(400, new { message = "Request body is required." });
                }

                var input = JsonSerializer.Deserialize<UpdateTaskDto>(request.Body, new JsonSerializerOptions
                {
                    PropertyNameCaseInsensitive = true
                });

                if (input == null)
                {
                    return JsonResponse(400, new { message = "Invalid body." });
                }

                var setClauses = new List<string>();
                if (input.Title != null) setClauses.Add("Title = @title");
                if (input.Description != null) setClauses.Add("Description = @description");
                if (input.Status != null) setClauses.Add("Status = @status");

                if (setClauses.Count == 0)
                {
                    return JsonResponse(400, new { message = "No fields to update." });
                }

                var sql = new StringBuilder();
                sql.Append("UPDATE Tasks SET ");
                sql.Append(string.Join(", ", setClauses));
                sql.Append(" WHERE Id = @id; ");
                sql.Append("SELECT Id, Title, Description, Status FROM Tasks WHERE Id = @id;");

                await using var connection = CreateConnectionFromEnv();
                await connection.OpenAsync();

                await using var cmd = new MySqlCommand(sql.ToString(), connection);
                cmd.Parameters.AddWithValue("@id", id.Value);
                if (input.Title != null) cmd.Parameters.AddWithValue("@title", input.Title);
                if (input.Description != null) cmd.Parameters.AddWithValue("@description", input.Description);
                if (input.Status != null) cmd.Parameters.AddWithValue("@status", input.Status);

                await using var reader = await cmd.ExecuteReaderAsync();
                // Move to second result set (SELECT)
                if (reader.NextResult() && await reader.ReadAsync())
                {
                    var task = new TaskEntity
                    {
                        Id = reader.GetInt32(0),
                        Title = reader.GetString(1),
                        Description = reader.IsDBNull(2) ? string.Empty : reader.GetString(2),
                        Status = reader.GetString(3)
                    };
                    return JsonResponse(200, task);
                }

                return JsonResponse(404, new { message = "Not found." });
            }
            catch (Exception ex)
            {
                context.Logger.LogError(ex.ToString());
                return JsonResponse(500, new { message = "Internal Server Error" });
            }
        }

        private static int? ResolveId(APIGatewayProxyRequest request)
        {
            if (request?.PathParameters != null && request.PathParameters.TryGetValue("id", out var idStr) && int.TryParse(idStr, out var idPath))
            {
                return idPath;
            }

            if (request?.QueryStringParameters != null && request.QueryStringParameters.TryGetValue("id", out var idQ) && int.TryParse(idQ, out var idQuery))
            {
                return idQuery;
            }

            return null;
        }

        private static MySqlConnection CreateConnectionFromEnv()
        {
            var host = Environment.GetEnvironmentVariable("DB_HOST");
            var port = Environment.GetEnvironmentVariable("DB_PORT") ?? "3306";
            var user = Environment.GetEnvironmentVariable("DB_USER");
            var password = Environment.GetEnvironmentVariable("DB_PASSWORD");
            var database = Environment.GetEnvironmentVariable("DB_NAME");

            var cs = $"Server={host};Port={port};Database={database};User ID={user};Password={password};SslMode=Preferred;";
            return new MySqlConnection(cs);
        }

        private static APIGatewayProxyResponse JsonResponse(int statusCode, object payload)
        {
            return new APIGatewayProxyResponse
            {
                StatusCode = statusCode,
                Body = JsonSerializer.Serialize(payload),
                Headers = new Dictionary<string, string>
                {
                    { "Content-Type", "application/json" },
                    { "Access-Control-Allow-Origin", "*" },
                    { "Access-Control-Allow-Methods", "GET,POST,PUT,DELETE,OPTIONS" },
                    { "Access-Control-Allow-Headers", "Content-Type,Authorization" }
                }
            };
        }
    }

    internal class UpdateTaskDto
    {
        public string? Title { get; set; }
        public string? Description { get; set; }
        public string? Status { get; set; }
    }

    internal class TaskEntity
    {
        public int Id { get; set; }
        public string Title { get; set; } = string.Empty;
        public string Description { get; set; } = string.Empty;
        public string Status { get; set; } = "To Do";
    }
}


